src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs

```diff
@@ -1567,17 +1572,56 @@ internal new NamedTypeSymbol GetSpecialType(SpecialType specialType)
         }
 
         /// <summary>
-        /// Cache of T to Nullable&lt;T&gt;.
+        /// Given a provided <paramref name="typeArgument"/>, gives back <see cref="Nullable{T}"/> constructed with that
+        /// argument.  This function is only intended to be used for very common instantiations produced heavily during
+        /// binding.  Specifically, the nullable versions of enums, and the nullable versions of core built-ins.  So
+        /// many of these are created that it's worthwhile to cache, keeping overall garbage low, while not ballooning
+        /// the size of the cache itself.
         /// </summary>
-        private ImmutableSegmentedDictionary<TypeSymbol, NamedTypeSymbol> _typeToNullableVersion = ImmutableSegmentedDictionary<TypeSymbol, NamedTypeSymbol>.Empty;
-
         internal NamedTypeSymbol GetOrCreateNullableType(TypeSymbol typeArgument)
-            => RoslynImmutableInterlocked.GetOrAdd(
+        {
+#if DEBUG
+            Debug.Assert(isSupportedType(typeArgument));
+#endif
+
+            return RoslynImmutableInterlocked.GetOrAdd(
                 ref _typeToNullableVersion,
                 typeArgument,
                 static (typeArgument, @this) => @this.GetSpecialType(SpecialType.System_Nullable_T).Construct(typeArgument),
                 this);
 
+#if DEBUG
+            static bool isSupportedType(TypeSymbol typeArgument)
+            {
+                if (typeArgument.IsEnumType())
```

Comment:
>typeArgument.IsEnumType() [](http://example.com/codeflow?start=20&length=25)

Why is it a good idea to cache for enum types? This is not a closed set of types.

Reply:
Because it's the largest source of constructing these types as all the work is done around enum operators and lifted operators.  It's not a closed set.  But they're also a tiny minority of all types, so caching does not produce much here, but saves a large amount of overall allocations.

Reply:
For context, enums are by far the heaviest source of construct allocations during typing: 

![image](https://user-images.githubusercontent.com/4564579/231871397-85253490-3b1c-480e-a655-7940c4480377.png)

Reply:
> Because it's the largest source of constructing these types as all the work is done around enum operators

Doesn't this depend mostly on the user code?

Reply:
Effectively, an external consumer, regardless of what is in the source code, can grow this cache unbounded.